### PR TITLE
bugfix: pouch pull failed

### LINKED ISSUE #93: [bug report] pouch pull failed
**Issue Description**
is pouch pull function fully completed? and is there any doc about pouch usage?
i got failed when using pouch pull.

**Describe what happened**:
1. i just use pouch pull nginx, and return an error.
image
2. i use the image from alibaba to do pouch pull, also got error.

> ./pouch pull reg.docker.alibaba-inc.com/busybox:latest

image
image
3. i use pouch create to create a container from image. i found that there is no image in "pouch images".
image

**Describe what you expected to happen**:
1. i want to got docker images from all kinds of repository.
2. i think pouch should pull images automatically when i use "pouch create ..."
3. also, pouch pull should be worked.

**How to reproduce it (as minimally and precisely as possible)**:
1. ./pouch pull nginx
2. ./pouch pull reg.docker.alibaba-inc.com/busybox:latest
3.

**Anything else we need to know?**:
image

**Environment**:
- pouch version (use `pouch version`):

> GoVersion:       go1.8.3
KernelVersion:
Os:              linux
Version:         0.1.0-dev
APIVersion:      vx.y.z
Arch:            amd64
BuildTime:       2017-11-14T10:46:18.071044934+08:00
GitCommit:

- OS (e.g. from /etc/os-release):

> NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

- Kernel (e.g. `uname -a`):

> Linux thinkcenter 3.10.0-693.2.2.el7.x86_64 #1 SMP Tue Sep 12 22:26:13 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

- Install tools:
- Others:

> runc version 1.0.0-rc4+dev
commit: b2567b37d7b75eb4cf325b77297b140ea686ce8f
spec: 1.0.0

containerd github.com/containerd/containerd v1.0.0-beta.3-8-g732b156 732b1560b495fa0a98d986aa57b365b6111ddae0

### LINKED ISSUE #123: Question about the images in pouch
**Issue Description**

Tried to run `pouch pull` to pull images just like docker does but failed.

**Describe what happened**:

Run `pouch pull hello-world:latest` but failed to get the image. I dived into the code and found that pouch will parse the URL `dummy://hello-world:latest` and get the object from the URL. I am not sure what `dummy://` is, and it seems that we can not use pouch like docker. Could you give me some guides to understand the concept of image in pouch?

I'd appreciate it.

#### Pouchd Log

```
INFO[2017-11-19 17:03:04.337774781] Calling POST /images/create?fromImage=hello-world&tag=latest, client @ 
ERRO[2017-11-19 17:03:04.337858143] failed to pull image hello-world:latest: failed to pull image: object required 
2017/11/19 17:03:04 http: multiple response.WriteHeader calls
ERRO[2017-11-19 17:03:04.337875015] invoke /images/create?fromImage=hello-world&tag=latest error failed to pull image: object required. client @ 
INFO[2017-11-19 17:03:04.438008556] fetch progress exited, ref: hello-world:latest. 
ERRO[2017-11-19 17:03:04.438060020] failed to write object: http: wrote more than the declared Content-Length 
```

#### Pouch Log

```
failed to decode: invalid character 'i' in literal false (expecting 'l') 
```

**Describe what you expected to happen**:

Pull the images successfully from docker hub or other registries.

**How to reproduce it (as minimally and precisely as possible)**:

1. `pouch pull hello-world:latest`

**Anything else we need to know?**:

Nothing.

**Environment**:

- pouch version (use `pouch version`):
```
BuildTime:       2017-11-19T16:57:07.198242422+08:00
GitCommit:       
GoVersion:       go1.9.1
KernelVersion:   
Os:              linux
Version:         0.1.0-dev
APIVersion:      vx.y.z
Arch:            amd64
```
- OS (e.g. from /etc/os-release):
```
NAME="Ubuntu"
VERSION="16.04.2 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.2 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
```
- Kernel (e.g. `uname -a`):
```
Linux ist 4.4.0-78-generic #99-Ubuntu SMP Thu Apr 27 15:29:09 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```
- Install tools:
```
make build
```

### PR DESCRIPTION
**Pull request by zhubingbing**
Signed-off-by: zhubingbing zhubingbing10@gmail.com

**1.Describe what this PR did**
fix pouch pull failed

1. pouch use default registry docker.io
2. 

**2.Does this pull request fix one issue?** 
fix #93 
fix #123

**3.Describe how you did it**
set default registry is docker.io

**4.Describe how to verify it**
```
[root@kolla-dev pouch]# ./pouch pull nginx
docker.io/library/nginx:latest: resolving      |--------------------------------------|
elapsed: 0.0 s                  total:   0.0 B (0.0 B/s)
^C
[root@kolla-dev pouch]# ./pouch pull zhubingbing/nginx
docker.io/zhubingbing/nginx:latest: resolving      |--------------------------------------|
elapsed: 0.0 s                      total:   0.0 B (0.0 B/s)
^C
[root@kolla-dev pouch]# ./pouch pull zhubingbing/nginx:v1
docker.io/zhubingbing/nginx:v1: resolving      |--------------------------------------|
elapsed: 0.0 s                  total:   0.0 B (0.0 B/s)
```

**5.Special notes for reviews**

### CODE REVIEW

**Comment by allencloud** [comment]
CI fails, please take a look at that. @zhubingbing 
In addition, we had better submit only one commit in one pull request. 
You can use `git commit -s --amend` to add details in the current commit without adding a new.

**Comment by allencloud** [comment]
Could you squash your commits into a single one and sign off your commits? @zhubingbing 

In addition, I think we implement this in cli client. But more ideally, it is server side to implement this.
Are you interested in implement this in a follow-up Pull Request?

**Comment by zhubingbing** [comment]
1. ok, I don't know where to implement the service. I don't see any code for the server。
   we can discuss  together 
2. for intergration test，i agree we shoud add integration test.  thanks @Letty5411   i try do it

**Comment by setekhid** [comment]
Have you guys considered about private or non-official registry?! It seems that the line 124 will trigger a bug with this image name "localhost:8080/foo/bar:latest".

One of you guys has already committed some similar code to my repo, so I know it.

**Comment by zhubingbing** [comment]
it can support private repo @setekhid

**Comment by allencloud** [comment]
Currently we will support private repo as soon as possible, and please help us review or test once we are ready. Thanks a lot in advance. @setekhid

File: cli/pull.go [hunk 1]
```diff
@@ -133,5 +133,14 @@ func parseNameTag(input string) (string, string) {
 		tag = fields[1]
 	}
 
+	if strings.Contains(name, "/") {
+		count := strings.Count(name, "/")
+		if count == 1 {
+			name = "docker.io/" + name
```
Comment (allencloud) [inline cli/pull.go]: Actually hard code is not our first choice. Maybe we can set a const variable there to represent this.

**Comment by allencloud** [comment]
any update? @zhubingbing

**Comment by janynedoe48** [comment]
On Dec 11, 2017 10:00 PM, "pouch_robot" <notifications@github.com> wrote:

ping @zhubingbing <https://github.com/zhubingbing>
Conflict happens after merging a previous commit.
Please rebase the branch against master and push it back again. Thanks a
lot.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/alibaba/pouch/pull/146#issuecomment-350953797>, or mute
the thread

**Comment by allencloud** [comment]
https://github.com/alibaba/pouch/pull/266 has fixed this. I am closing this.

**Pull request closed without merging**